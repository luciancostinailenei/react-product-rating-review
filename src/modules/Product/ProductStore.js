import { action, computed, observable } from 'mobx'

import moment from 'moment'

class ProductStore {
  @observable reviews = []
  @observable isProductReviewFormDisplayed = false

  @action
  addReview = review => {
    this.reviews.push(review)
  }

  @computed
  get reviewsCount() {
    return this.reviews.length
  }

  @computed
  get sortedReviews() {
    const sortedReviews = this.reviews.sort((a, b) => {
      if (moment(a.date) > moment(b.date)) {
        return -1
      }

      if (moment(a.date) < moment(b.date)) {
        return 1
      }

      return 0
    })

    return sortedReviews
  }

  @computed
  get averageScore() {
    const numberOfReviews = this.reviewsCount

    if (numberOfReviews === 0) {
      return 0
    }

    const scoreSum = this.reviews.reduce((acc, review) => {
      return acc + review.rating
    }, 0)

    const averageScore = scoreSum / numberOfReviews
    return averageScore % 1 === 0 ? averageScore : averageScore.toFixed(1)
  }
}

const store = new ProductStore()
export default store
