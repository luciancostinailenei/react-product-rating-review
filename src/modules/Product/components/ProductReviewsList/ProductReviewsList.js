import React, { Component } from 'react'
import { inject, observer } from 'mobx-react'

import './ProductReviewsList.scss'

import ProductReview from '../ProductReview/ProductReview'

@inject('ProductStore')
@observer
export default class ProductReviewsList extends Component {
  render() {
    return (
      <div className="ProductReviewsList">
        <div className="ProductReviewsList__reviews-overview">
          <p className="ProductReviewsList__general-rating">
            General Product Rating:
            <span className="ProductReviewsList__average-score">
              {this.props.ProductStore.averageScore}
            </span>
          </p>

          <p className="ProductReviewsList__reviews-number">
            <b> {this.props.ProductStore.reviewsCount} </b> reviews
          </p>
        </div>

        <ul className="ProductReviewsList__reviews-list list list--no-bullets list--no-padding-left">
          {this.props.ProductStore.sortedReviews.map((review, index) => {
            return (
              <ProductReview reviewData={review} key={index} index={index} />
            )
          })}
        </ul>
      </div>
    )
  }
}
