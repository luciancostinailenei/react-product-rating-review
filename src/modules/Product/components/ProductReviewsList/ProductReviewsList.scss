.ProductReviewsList { padding: 10px; }

    .ProductReviewsList__general-rating {
        color: var(--teal-500);
        display: flex;
        font-size: 18px;
        margin-bottom: 0;
    }

    .ProductReviewsList__average-score {
        color: var(--amber-700);
        font-weight: bold;
        font-size: 20px;
        margin-left: 10px;
    }

    .ProductReviewsList__reviews-number {
        margin-top: 0;
        margin-left: 3px;
    }

    .ProductReviewsList__reviews-list { margin-top: 40px; }