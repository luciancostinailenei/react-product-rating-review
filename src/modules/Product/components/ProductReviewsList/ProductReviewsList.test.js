import { configure, shallow } from 'enzyme'
import React from 'react'
import Adapter from 'enzyme-adapter-react-16'
import { Provider } from 'mobx-react'
import 'jsdom-global/register'
import toJson, { createSerializer } from 'enzyme-to-json'

import ProductReviewsList from './ProductReviewsList'

import ProductStore from '../../ProductStore'

expect.addSnapshotSerializer(createSerializer({ mode: 'deep' }))

configure({ adapter: new Adapter() })

describe('ProductReviewsList', () => {
  it('matches snapshot', () => {
    const ProductReviewsListWrapper = shallow(
      <ProductReviewsList ProductStore={ProductStore} />
    )

    expect(toJson(ProductReviewsListWrapper)).toMatchSnapshot()
  })

  it('reacts to store changes', function() {
    ProductStore.addReview({
      rating: 5,
      title: 'Test Review Title 1',
      text: 'Review Text Test 1',
      user: 'Lucian',
      date: new Date()
    })

    const ProductReviewsListWrapper = shallow(
      <ProductReviewsList ProductStore={ProductStore} />
    )

    expect(
      ProductReviewsListWrapper.dive()
        .find('.ProductReviewsList__average-score')
        .text()
    ).toBe('5')

    expect(
      ProductReviewsListWrapper.dive()
        .find('.ProductReviewsList__reviews-number b')
        .text()
    ).toBe('1')
  })
})
