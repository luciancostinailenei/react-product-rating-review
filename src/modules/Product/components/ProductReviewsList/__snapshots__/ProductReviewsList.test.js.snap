// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductReviewsList matches snapshot 1`] = `
<ProductReviewsList
  ProductStore={ProductStore {}}
/>
`;
